isis/src/base/apps/hist/main.cpp

```diff
@@ -27,16 +27,38 @@ void IsisMain() {
     QString msg = "The [TO] parameter must be entered";
     throw IException(IException::User, msg, _FILEINFO_);
   }
-  
-  // Setup the histogram
-  Histogram hist(*icube, 1, p.Progress() );
-  if(ui.WasEntered("MINIMUM") ) {
-    hist.SetValidRange(ui.GetDouble("MINIMUM"),ui.GetDouble("MAXIMUM") );     
-  }
 
-  if(ui.WasEntered("NBINS")) {
-    hist.SetBins(ui.GetInteger("NBINS"));
+  Histogram *hist;
+  if (ui.WasEntered("MINIMUM") && ui.WasEntered("MAXIMUM")){
+    int nbins = 0;
+
+    if (ui.WasEntered("NBINS")){
+      nbins = ui.GetInteger("NBINS");
+    }
+    else {
+      // Calculate bins based on data type.
+      // Bin calculations are based on logic in Histogram::InitializeFromCube
+      if (icube->pixelType() == UnsignedByte) {
+        nbins = 256;
+      }
+      else if (icube->pixelType() == SignedWord ||
+               icube->pixelType() == UnsignedWord ||
+               icube->pixelType() == UnsignedInteger ||
+               icube->pixelType() == SignedInteger ||
+               icube->pixelType() == Real) {
+        nbins = 65536;
+      }
+      else {
+        IString msg = "Unsupported pixel type";
+        throw IException(IException::Programmer, msg, _FILEINFO_);
+      }
+    }
+    hist = new Histogram(ui.GetDouble("MINIMUM"), ui.GetDouble("MAXIMUM"), nbins);
   }
+  else {
+    hist = new Histogram(*icube, 1, p.Progress());
```

Comment:
The Histogram needs to be deleted at the end of the function otherwise it's a memory leak.